slackapi/node-slack-sdk
# web-api: public canvas APIs

Adds several new Canvas APIs: create, edit, delete, sections.lookup, access.delete, access.set and conversations.canvases.create.

The response shapes were generated using https://github.com/slackapi/java-slack-sdk/pull/1322

# Manual Testing

To give this PR a try, check out this branch locally, then you can add a `test.ts` to the `packages/web-api` directory that is the following code. It exercises the various Canvas APIs: creates a canvas, sets access on it, looks up sections on it, edits it, and finally deletes it. You will need to export a `SLACK_BOT_TOKEN` environment variable before running it, that is a bot token with at least `canvases:write`, `canvases:read` and `files:read` scopes. This script code executes against the dev environment. Finally, make sure to adjust the channel and user IDs at the top of the script to match a channel and user ID in the workspace you are testing in.

You can run it with `npx tsx test.ts`:

```typescript
import assert from 'assert';
import { WebClient, LogLevel } from './src/';
(async () => {
  const channel = 'C0147RBUF19';
  const user = 'U0147RBMVB5';
  const client = new WebClient(process.env.SLACK_BOT_TOKEN, {
    logLevel: LogLevel.DEBUG,
    slackApiUrl: 'https://dev.slack.com/api/',
  });
  const canvas = await client.canvases.create({"title":"new canvas" + new Date(), "document_content": {"type":"markdown","markdown":`
# Heading ONE

> And this is a quote

## HEading TWO

- list one
- list two
`}});
  console.log('create resp', canvas);
  const info = await client.files.info({
    file: canvas.canvas_id!,
  });
  const permalink = info.file?.permalink;
  await client.chat.postMessage({
    channel,
    text: `Here's the canvas I just created: ${permalink}`
  });
  console.log('canvases.access.set resp', await client.canvases.access.set({
    canvas_id: canvas.canvas_id!,
    access_level: 'write',
    channel_ids: [channel],
  }));
  console.log('canvases.access.set resp', await client.canvases.access.set({
    canvas_id: canvas.canvas_id!,
    access_level: 'write',
    user_ids: [user],
  }));
  console.log('canvases.access.delete resp', await client.canvases.access.delete({
    canvas_id: canvas.canvas_id!,
    user_ids: [user],
  }));
  console.log('canvases.access.delete resp', await client.canvases.access.delete({
    canvas_id: canvas.canvas_id!,
    channel_ids: [channel],
  }));
  const { sections } = await client.canvases.sections.lookup({
    canvas_id: canvas.canvas_id!,
    criteria: {
      contains_text: "list"
    }
  });
  const originalNumberOfListItems = sections!.length;
  console.log('canvases.edit resp', await client.canvases.edit({
    canvas_id: canvas.canvas_id!,
    changes: [
      {
        operation: 'insert_after',
        section_id: sections![0].id!,
        document_content: {
          type: 'markdown',
          markdown: `- a new list item appears!\n`
        }
      }
    ]
  }));
  const { sections: sectionsTwo } = await client.canvases.sections.lookup({
    canvas_id: canvas.canvas_id!,
    criteria: {
      contains_text: "list"
    }
  });
  assert.equal(sectionsTwo!.length, originalNumberOfListItems + 1, 'expected one more section retrieved from lookup that contain the word "list" than originally created')
  console.log('canvases.delete resp', await client.canvases.delete({
    canvas_id: canvas.canvas_id!,
  }));
})();
```
